h2o-core/src/main/java/hex/grid/HyperSpaceWalker.java

```diff
@@ -191,22 +195,30 @@ public BaseWalker(MP params,
       _hyperParams = hyperParams;
       _paramsBuilderFactory = paramsBuilderFactory;
       _hyperParamNames = hyperParams.keySet().toArray(new String[0]);
-      _maxHyperSpaceSize = computeMaxSizeOfHyperSpace();
+      _hyperParamSubspaces = extractSubspaces();
+      _hyperParamNamesSubspace = extractSubspaceNames();
+      if(_hyperParams.get(SUBSPACES) != null) { _hyperParams.remove(SUBSPACES); }
       _search_criteria = search_criteria;
-
+      _maxHyperSpaceSize = computeMaxSizeOfHyperSpace();
+      
       // Sanity check the hyperParams map, and check it against the params object
       try {
         _defaultParams = (MP) params.getClass().newInstance();
       } catch (Exception e) {
         throw new H2OIllegalArgumentException("Failed to instantiate a new Model.Parameters object to get the default values.");
       }
-      validateParams();
+      validateParams(_hyperParams, false);
+      Arrays.stream(_hyperParamSubspaces).forEach(subspace -> validateParams(subspace, true));
     } // BaseWalker()
 
     @Override
     public String[] getHyperParamNames() {
       return _hyperParamNames;
     }
+    
+    public String[] getHyperParamNamesSubspace() {
+      return _hyperParamNamesSubspace;
+    }
```

Comment:
for clarity, I would rename this method and attribute to 
`getAllHyperParamNamesInSubspaces` / `_allHyperParamNamesInSubspaces`